src/needless_continue.rs

```diff
@@ -0,0 +1,547 @@
+        ExprLoop(ref loop_block, _)     => {
+            for stmt in &loop_block.stmts {
+                match stmt.node {
+                    StmtSemi(ref if_expr, _) | StmtExpr(ref if_expr, _) => {
+                        found = check_for_continue_in_else(ctx, expr, if_expr, DROP_ELSE_BLOCK_AND_MERGE_MSG);
+                        if found { break; }
+                        found = check_for_continue_in_if(ctx, expr, if_expr, REWRITE_IF_COND_MSG);
+                        if found { break; }
+                    },
+                    _ => { },
+                }
+            }
+            if !found {
+                if let Some(ref if_expr) = loop_block.expr {
+                    found = check_for_continue_in_else(ctx, expr, if_expr, DROP_ELSE_BLOCK_MSG);
+                    if !found {
+                        check_for_continue_in_if(ctx, expr, if_expr, REWRITE_IF_COND_MSG); 
+                    }
+                }
+            }
+        },
+        _ => { }
+    }
+}
+
+/// Eats at `s` from the end till a closing brace `}` is encountered, and then
+/// continues eating till a non-whitespace character is found.
+/// e.g., the string (using Python style triple quotes for clarity)
+///
+/// """
+/// {
+///     let x = 5;
+/// }
+/// """
+///
+/// is transformed to 
+///
+/// """
+/// {
+///     let x = 5;"""
+///
+fn erode_from_back(s: &str) -> String {
+    let mut ret = String::from(s); 
+    while ret.pop().map_or(false, |c| c != '}') { }
+    while let Some(c) = ret.pop() {
+        if !c.is_whitespace() {
+            ret.push(c);
+            break;
+        }
+    }
+    ret
+}
+
+/// Constructs a code suggestion for an `if` expression within a loop
+/// that has an else arm that just contains a `continue`.
+/// 
+/// If there is no code following the if expr in the loop, we suggest that
+/// the user simply drop the else clause. If there is code after the if expr,
+/// we suggest a merger of this trailing code with the block of the if expr.
+/// e.g., 
+///     
+///     loop {
+///         // region A
+///         if cond {
+///             // region B
+///         } else {
+///             continue;
+///         }
+///         // region C
+///     }
+/// 
+/// will yield a suggestion like
+/// 
+///     loop {
+///         // region A
+///         if cond {
+///             // region B
+///             // region C
+///         }
+///     }
+///
+/// Regions B and C are merged together in the if block.
+fn suggestion_snippet_for_continue_inside_else(ctx: &LateContext,
+                                               loop_expr: &Expr,
+                                               if_expr: &Expr,
+                                               if_block: &Block,
+                                               header: &str) -> String {
+
+    let if_code = erode_from_back(&trim_indent(&snippet(
+        ctx,
+        Span { lo: if_expr.span.lo, hi: if_block.span.hi, expn_id: if_expr.span.expn_id },
```

Comment:
I'm really wary of constructing spans by hand. When macros are involved this can get ugly (if the expansion ids are different).

@llogiq What do you think of this?